src/entities/priceImpact/index.ts

```diff
@@ -45,4 +54,134 @@ export class PriceImpact {
+    ): Promise<PriceImpactAmount> => {
+        // inputs are being validated within AddLiquidity
+
+        // simulate adding liquidity to get amounts in
+        const addLiquidity = new AddLiquidity();
+        const { amountsIn, bptOut } = await addLiquidity.query(
+            input,
+            poolState,
+        );
+        const poolTokens = amountsIn.map((a) => a.token);
+
+        // simulate removing liquidity to get amounts out
+        const removeLiquidity = new RemoveLiquidity();
+        const removeLiquidityInput: RemoveLiquidityInput = {
+            chainId: input.chainId,
+            rpcUrl: input.rpcUrl,
+            bptIn: bptOut.toInputAmount(),
+            kind: RemoveLiquidityKind.Proportional,
+        };
+        const { amountsOut } = await removeLiquidity.query(
+            removeLiquidityInput,
+            poolState,
+        );
+
+        // diffs between unbalanced and proportional amounts
+        const diffs = amountsOut.map((a, i) => a.amount - amountsIn[i].amount);
+
+        // get how much BPT each diff would mint
+        const diffBPTs: bigint[] = [];
+        for (let i = 0; i < diffs.length; i++) {
+            if (diffs[i] === 0n) {
+                diffBPTs.push(0n);
+            } else {
+                diffBPTs.push(await queryBPTForDiffAtIndex(i));
+            }
+        }
+
+        // zero out diffs by swapping between tokens from proportionalAmounts
+        // to exactAmountsIn, leaving the remaining diff within a single token
+        const remainingDiffIndex = await zeroOutDiffs(diffs, diffBPTs);
+
+        // get relevant amounts for price impact calculation
+        const amountInitial = parseFloat(
+            formatUnits(
+                amountsIn[remainingDiffIndex].amount,
+                amountsIn[remainingDiffIndex].token.decimals,
+            ),
+        );
+        const amountDiff = parseFloat(
+            formatUnits(
+                abs(diffs[remainingDiffIndex]),
+                amountsIn[remainingDiffIndex].token.decimals,
+            ),
+        );
+
+        // calculate price impact using ABA method
+        const priceImpact = amountDiff / amountInitial / 2;
+        return PriceImpactAmount.fromDecimal(`${priceImpact}`);
+
+        // helper functions
+
+        async function zeroOutDiffs(diffs: bigint[], diffBPTs: bigint[]) {
+            let minNegativeDiffIndex = 0;
+            const nonZeroDiffs = diffs.filter((d) => d !== 0n);
+            for (let i = 0; i < nonZeroDiffs.length - 1; i++) {
+                const minPositiveDiffIndex = diffBPTs.findIndex(
+                    (diffBPT) =>
+                        diffBPT === min(diffBPTs.filter((a) => a > 0n)),
+                );
+                minNegativeDiffIndex = diffBPTs.findIndex(
+                    (diffBPT) =>
+                        diffBPT === max(diffBPTs.filter((a) => a < 0n)),
+                );
+
+                let kind: SwapKind;
+                let givenIndex: number;
```

Comment:
I agree that adding `token` to the name is helpful 👍 
But I'm not so sure about `received`, because I think it may be confused with `tokenOut`, which is not always the case. For `GivenOut` swaps, `received` would be related to `tokenIn`, which is a bit odd.
I decided to go with `result` because that's what is returned from the `query` action [here](https://github.com/balancer/b-sdk/blob/d0a871cd59855f156f249dab218b9ed175e39cc2/src/entities/utils/doQuerySwap.ts#L46)
With that in mind, I think we should go with `givenTokenIndex` and `resultTokenIndex`.
Let me know if that makes sense.